src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+
+    def _add_wire_labels(self):
+        """
+        Add wire labels to the circuit
+        """
+
+        default_labels = [f"q{i}" for i in range(self._qwires)] + [
+            f"c{i}" for i in range(self._cwires)
+        ]
+        for i, label in enumerate(default_labels):
+            self._render_strs["mid_frame"][i] = (
+                f" {label} :" + self._render_strs["mid_frame"][i]
+            )
+            self._render_strs["top_frame"][i] = " " * (len(f" {label} :") + 2)
+            self._render_strs["bot_frame"][i] = " " * (len(f" {label} :") + 2)
+
+        self._layer_list = {
+            i: [len(self._render_strs["mid_frame"][i])]
+            for i in range(self._qwires + self._cwires)
+        }
+
+    def _extend_line(self):
+        """
+        Extend all the wires to the same length
+        """
+
+        wire_len = [sum(wire) for wire in self._layer_list.values()]
+        max_len = max(wire_len) + self.style.end_wire_ext
+
+        for i, length in enumerate(wire_len):
+            if length < max_len:
+                diff = max_len - length
+                if i < self._qwires:
+                    self._render_strs["mid_frame"][i] += "─" * diff
+                else:
+                    self._render_strs["mid_frame"][i] += "═" * diff
+
+    def _draw_singleq_gate(self, gate_name: str):
+        """
+        Draw a single qubit gate
+
+        Parameters
+        ----------
+        gate : Gate
+            The gate to be drawn.
+
+        Returns
+        -------
+        tuple
+            The parts of the gate to be drawn. The parts are the top_frame, mid_frame, and bot_frame.
+
+        int
+            The width of the gate.
+        """
+
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_name))
+        pad = " " * ceil(self.style.gate_pad)
+
+        top_frame = f" ┌{lid_seg}┐ "
+        mid_frame = f"─┤{pad}{gate_name}{pad}├─"
+        bot_frame = f" └{lid_seg}┘ "
+
+        return (top_frame, mid_frame, bot_frame), len(top_frame)
+
+    def _draw_multiq_gate(self, gate: Gate, gate_text: str):
+        """
+        Draw a multi qubit gate
+
+        Parameters
+        ----------
+        gate : Gate
+            The gate to be drawn.
+
+        Returns
+        -------
+        tuple
+            The parts of the gate to be drawn.
+            i.e. the top_frame, mid_frame, mid_connect, mid_connect_label, and bot_frame.
+
+        int
+            The width of the gate.
+        """
+
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_text))
+        pad = " " * ceil(self.style.gate_pad)
+
+        top_frame = f" ┌{lid_seg}┐ "
+        bot_frame = f" └{lid_seg}┘ "
+        mid_frame = f" │{pad}{' ' * len(gate_text)}{pad}│ "
+        mid_connect = f"─┤{pad}{' ' * len(gate_text)}{pad}├─"
+        mid_connect_label = f"─┤{pad}{gate_text}{pad}├─"
```

Comment:
Low priority:
I feel like ` _draw_singleq_gate` can be merged into this `_draw_multiq_gate`.

Reply:
Sure, the `_draw_singleq_gate` function follows a similar structure to MatRenderer, with separate drawing functions for single and multi-qubit gates.